toolchains/symbiflow.py

```diff
@@ -580,177 +583,223 @@ def check_env():
         }
 
 
-class NextpnrXilinx(Toolchain):
+class NextpnrGeneric(Toolchain):
     '''nextpnr using Yosys for synthesis'''
-    carries = (False, )
+    resources_map = {
+        'xilinx':
+            {
+                'LUT': ('SLICE_LUTX', ),
+                'DFF': ('SLICE_FFX', ),
+                'CARRY': ('CARRY4', ),
+                'IOB':
+                    (
+                        'IOB33M_OUTBUF',
+                        'IOB33M_INBUF_EN',
+                        'IOB33_OUTBUF',
+                        'IOB33_INBUF_EN',
+                    ),
+                'PLL': ('PLLE2_ADV_PLLE2_ADV', ),
+                'BRAM': (
+                    'RAMB18E1_RAMB18E1',
+                    'RAMB36E1_RAMB36E1',
+                ),
+            },
+        'fpga_interchange':
+            {
+                'xilinx':
+                    {
+                        'LUT': ('LUTS', ),
+                        'DFF': ('CARRY', ),
+                        'CARRY': ('FLIP_FLOPS', ),
+                        'IOB': (
+                            'IBUFs',
+                            'OBUFs',
+                        ),
+                        'PLL': ('PLL', ),
+                        'BRAM': ('BRAMS', ),
+                    },
+            },
+    }
 
     def __init__(self, rootdir):
         Toolchain.__init__(self, rootdir)
-        self.toolchain = 'nextpnr-xilinx'
-        self.files = []
-        self.fasm2bels = False
+        self.arch = None
+        self.vendor = None
+        self.chipdb = None
+        self.chip = None
+        self.builddir = "."
         self.dbroot = None
+        self.files = list()
+        self.yosys_synth_opts = list()
+        self.yosys_template = None
+        self.schema_dir = None
+        self.device_file = None
+        self.options = str()
+        self.fasm2bels = False
 
-    def run_steps(self):
-        with Timed(self, 'bitstream'):
-            self.backend.build_main(self.project_name + '.bit')
-
-    def run(self):
-        with Timed(self, 'total'):
-            with Timed(self, 'prepare'):
-                os.makedirs(self.out_dir, exist_ok=True)
+    def get_share_data(self):
+        out = subprocess.run(
+            ["find", ".", "-name", self.toolchain_bin], stdout=subprocess.PIPE
+        )
+        nextpnr_locations = out.stdout.decode('utf-8').split('\n')
+        nextpnr_location = None
+
+        for location in nextpnr_locations:
+            if "env/bin/{}".format(self.toolchain_bin) in location:
```

Comment:
the `env/bin` assumes that there is a special environment (e.g. conda in this case) that nextpnr is installed into. Might be worth removing the `env`?